swan-lake/development-tutorials/test-document-the-code/ballerina-profiler.md

```diff
@@ -0,0 +1,365 @@
+final Country[] & readonly countries;
+
+function init() returns error? {
+    countries = check diseaseEp->/covid\-19/countries;
+}
+
+service /covid19/countries on new http:Listener(8080) {
+    resource function get summary() returns json {
+        Data[] listResult = from var {country, continent, population, cases, deaths} in countries
+            where hasSignificantPopulation(population, deaths)
+            let decimal caseFatalityRatio = <decimal>deaths / <decimal>cases * 100
+            order by caseFatalityRatio descending
+            limit 1000
+            select {country, continent, population, caseFatalityRatio};
+        return getTopSortedValues(listResult);
+    }
+
+    isolated resource function get names() returns json {
+        return from var {country, population, deaths} in countries
+            where hasNonZeroPopulation(population, deaths)
+            select {country};
+    }
+}
+
+isolated function getTopSortedValues(Data[] listResult) returns json[] {
+    json[] sortedArray = listResult.sort("ascending", getKey);
+    return getTopElements(sortedArray);
+}
+
+isolated function getTopElements(json[] sortedArray) returns json[] {
+    return from var i in sortedArray
+        limit 10
+        select i;
+}
+
+isolated function hasSignificantPopulation(int population, int deaths) returns boolean {
+    return population >= 100 && deaths >= 10;
+}
+
+isolated function hasNonZeroPopulation(int population, int deaths) returns boolean {
+    return population >= 0 && deaths >= 0;
+}
+
+isolated function getKey(record {|string country; string continent; int population; decimal caseFatalityRatio;|} recordVal) returns string {
+    return recordVal.country;
+}
+```
+
+Run the CLI command `bal profile` in the root directory to run the Ballerina package and profile it. This will run the service while profiling it.
+
+```
+$ bal profile
+Compiling source
+        profiler_demo/covid19_stats:0.1.0
+
+Generating executable
+        target/bin/covid19_stats.jar
+
+================================================================================
+Ballerina Profiler: Profiling...
+================================================================================
+Note: This is an experimental feature, which supports only a limited set of functionality.
+[1/6] Initializing...
+[2/6] Copying executable...
+[3/6] Performing analysis...
+[4/6] Instrumenting functions...
+ ○ Instrumented module count: 44
+ ○ Instrumented function count: 2935
+[5/6] Running executable...
+```
+
+Send 10 requests to each resource endpoint using the `curl` command.
+
+```
+$ curl localhost:8080/covid19/countries/names
+```
+
+```
+$ curl localhost:8080/covid19/countries/summary
+```
+
+Stop the service by sending the `SIGINT` signal to the Ballerina program. We can use `Ctrl+C` in the service running terminal to send the signal to it. This will stop both the Ballerina program and profiling it and will generate the profiler output.
```

Comment:
```suggestion
Stop the service by sending the `SIGINT` signal to the Ballerina program. You can use `Ctrl+C` in the service running terminal to send the signal to it. This will stop both the Ballerina program and profiling it and will generate the profiler output.
```